man/mopen.1
.\" 
.\" mopen manual page.
.\" (C) 2007 Novell, Inc. 
.\" Author:
.\"   Miguel de Icaza (miguel@gnu.org)
.\"
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.TH mopen "Mono 1.2, Moonlight 1.0"
.SH NAME
mopen, mopen1 \- Application launcher (including Mono, Moonlight and Silverlight).
.SH SYNOPSIS
.PP
.B mopen [options] file [arguments...]
.SH DESCRIPTION
The \fImopen\fP command is a tool to open applications, XAML files or
packages that contains executables, images and other resources in a
directory, it is designed to launch Silverligh 2.0 applications.   For
testing only 1.0 applications use the \fImopen1\fP command.
.PP
In addition to launching these applications, it is possible to run
multiple applications using the same Mono virtual machine.  This is
useful for example if you want to run a collection of small
applications, gadgets or widgets and do not want to pay the price of
memory consumption of running one entire virtual machine for every
application.  This is achieved by using Mono Application Domains: the
equivalent to processes within a Mono virtual machine, and it is the
same infrastructure used by ASP.NET applications to run multiple
isolated applications in the same virtual machine.
.PP
In its simplest incarnation, you can open a XAP file like this:
.nf

	$ mopen /opt/myapp/test.xap

.fi
This will load the test.xap file which is a self-contained Silverlight
application.    
.PP
It is possible to also load 1.0-style Silverlight applications
by passing it a XAML file and using the \fImopen1\FP command instead:
.nf
	$ mopen1 /opt/myapp/test.xaml
.fi
This will laod the xaml file and all of the dependencies it might have
from the directory where test.xaml is located.
.PP
mopen also supports launching these applications if they are contained
in a directory thta is on the PATH and such directory contains a file
called "default.xaml" or "main.exe".
.PP
In the case of applications that start up from XAML, the default
size of the window will be the one specified for the Canvas unless the
--geometry option has been used
.PP
With this configuration it is possible to have entire applications
that are packaged in a directory and can be deployed and copied by
copying the entire directory (drag and dropping the directory, using
cp -a, or any other mechanism to copy a directory).  
.SH OPTIONS
The following options are available:
.TP
.I "--debug"
Passes the flag --debug to the Mono runtime, enabling a line numbers
and a couple of other debugging features.
.TP
.I "--gdb"
This starts up Mono with gdb and prints out the command that you
should use to debug mopen.   This is useful when debugging XAML files
that cause problems with Moonlight.
.TP
.I "--desklet", "-d"
Turns on desklet mode.   In this mode the window decoration is not
shown and the default rendering mode will enable alpha transparency.
This means that anything painted with an alpha channel will be
properly transparent.
.TP
.I "-f", "--fixed"
Makes the toplevel window fixed, prevents the user from resizing it.
.TP
.I "--geometry=GEO", "-g GEO"
Sets the geometry of the window to be GEO, the format for this is
WIDTHxHEIGH where WIDTH and HEIGHT are integers, for example:
.nf

	mopen -g 1024x768

.fi
.TP
.I "--host NAME"
Loads the application into the host identified by NAME.   If no host
is specified this application is loaded into its own virtual machine
and can not be shared with others.
.TP
.I "--stories", "-s"
Triggers one by one all the named Storyboards in the given XAML file
when you click on the canvas surface. 
.TP
.I "--parseonly", "--parse-only"
Parse the XAML file but does not open a window. Useful for testing.
.TP
.I "--story=N1[,N2]"
Runs the storyboard named N1 when the mouse is clicked (this is the
x:Name property on the XAML file).  The storyboard playback will cycle
if you keep clicking.   You can use it like this:
.nf
.fi
	mopen --story FadeIn demo.xaml
	mopen --story FadeIn,FadeOut demo.xaml
	mopen --story FadeIn --story FadeOut demo.xaml
.TP
.I "--transparent", "-t"
Enables the alpha channel on the rendering surface.   Any areas that
are rendered with non-opaque values will be properly composited
against the underlying background by the system compositing manager.
.TP
.I "--timeout=X"
Close the mopen window after X seconds. Useful for testing.
.SH SHARING A VIRTUAL MACHINE
.PP
To load multiple independent applications into a single virtual
machine (the virtual machine will isolate each application from each
other using Application Domains, or AppDomains) you use the --host
option with the name of the domain where you want to load your
application.
.PP
For example, if you wanted to load various desktop gadgets on a single
process, you would load them like this:
.nf

	mopen --host desktop calendar
	mopen --host desktop stocktracker
	mopen --host desktop flicker-best-images
	mopen --host desktop cute-clock

.fi
You could load a separate set of programs on another VM, for example
while developing code that might be unstable:
.nf

	mopen --host devel mytestapp

.fi
.PP
To load a desklet, you would use:
.nf

	mopen --desklet mydesklet

.fi
.SH FILES
When a file is specified, if it exists on the current directory, that
one is loaded, otherwise directories on the PATH are searched and if
they contain the file 
.I default.xaml
or the file
.I main.exe
Those are loaded and launched
.SH MAILING LISTS
Mailing lists are listed at the
http://www.mono-project.com/Mailing_Lists
.PP
In addition Moonlight is discussed in the "Olive" group:
http://groups.google.com/group/mono-olive
.SH WEB SITE
http://www.mono-project.com
.SH SEE ALSO
.PP
mono(1),svg2xaml(1)


NEWS
Changes from 0.6 to 0.7
=======================

 * Plugin

   * Webkit loads the plugin (kangaroo, lewing)

   * The stream/downloader/request/response logic (used for
     downloading media) has been been almost entirely moved from the
     browser bridges into libmoon, with the browsers providing
     subclasses. (kangaroo, sde)

   * Finally add argument checking to all wrapped plugin objects
     (fejj).

   * Windowless mode fixes (lewing, toshok)

   * Plugin event handling fixes (lewing)


 * Engine

   * Many clock/animation framework fixes.  We now pass both animation
     matrix tests, and many, *many* other bugs (and regressions) have
     been fixed. (mdk).

   * Bug fixes in the Stroke{Collection}.HitTest and
     Stroke{Collection}.Bounds code (toshok, sde).

   * Namescope merging fixes (sde, jackson)

   * Parser fixes, and changes paving the way for 2.0 work (jackson)

   * Fix mouse event bubbling behavior (toshok)


 * Media

   * Big, big strides in our media framework and the various (file,
     http, mms) downloaders, (fejj, rolf, kangaroo, fer)

   * MMS stream selection (kangaroo)


 * Performance

   * Shape caching and bounds computation reduction (spouliot)

   * Geometry bounds work (spouliot)

   * Fast path for position updates (Canvas.Left/Canvas.Top) (toshok)

   * Improved temporary cairo surface bounds (lewing)

   * Glyph rendering speedups (fejj)

   * Resort by ZIndex as a dirty pass (toshok)


 * Silverlight 2.0

   * work is progressing. A very simple 2.0 application successfully
     ran. (miguel, jackson, sde).



tools/mopen/mopen.cs
//
// mopen.cs: Mono Open Application
//
// Opens a XAML file or an application
//
// Author:
//   Miguel de Icaza (miguel@novell.com)
//
//
// See LICENSE file in the Moonlight distribution for licensing details
//
// TODO:
//    * Implement --host
//    * Implement loading of exes
//    * Implement --fixed
//    * Make it so we can open directories that contain DIR/default.xaml
//      and setup the managed loader to default to this location to load
//      images from. 
//
// The idea is that we can write XAML-based applications and launch them with
// this thing, typing:
//
//    $ mopen calendar
//
// Would check if calendar/default.xaml exists, and if it does, configure the
// downloader to default to the directory calendar for obtaining images,
// resources and so on.
//
// If no default.xaml exists, we could check in order: a file with a .dll
// extension, if found, load it;   a file with any of the scripting languages
// extensions that the DLR knows about ".py", ".rb", ".js", ".vb" and load
// the proper compiler to run the code.
//
// The idea behind --host is to load multiple programs in a single instance of
// mono/moonlight using separate appdomains, so multiple applications (or
// desklets) can be loaded into the same process, saving resources.
//
// We could either use DBus# for this, or a simpler mechanism might be to use
// the named features from System.Threading to do this
//

using System;
using System.Collections.Generic;
using GLib;
using Gtk;
using Cairo;
using Gtk.Moonlight;
using System.Windows;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Controls;
using System.Windows.Input;
using System.IO;
using System.Collections;
using Mono;
using System.Reflection;
using System.Xml;
using Application = Gtk.Application;
using NDesk.Options;

class MonoOpen {
	static Window window;
	static bool fixedwindow = false;
	static int width = -1;
	static int height = -1;
	static bool transparent = false;	
	static bool desklet = false;
	static bool parse_only = false;
	static List<string> story_names = null;
	static List<Storyboard> storyboards = null;
	static int current_storyboard = 0;
	static bool all_stories = false;
	static bool sync = false;
	static int timeout = -1;
	
	static void ShowHelp (OptionSet o)
	{
		Console.WriteLine ("Usage is: mopen [args] [file.xaml|file.xap|dirname]\n\n" +
				   "Arguments are:\n");
		o.WriteOptionDescriptions (Console.Out);
	}

	[GLib.ConnectBefore]
	static void HandleExposeEvent (object sender, ExposeEventArgs expose_args)
	{
		 Widget w = (Widget)sender;
	    
		 using (Cairo.Context ctx = CompositeHelper.Create (w.GdkWindow)){
		 	ctx.Operator = Cairo.Operator.Source;
		 	ctx.Color = new Cairo.Color (1.0, 1.0, 1.0, 0.0);
		 	CompositeHelper.Region (ctx, expose_args.Event.Region);
		 	ctx.Fill ();
		}
	}

	static bool can_dragging = false;

	static void HandleMouseLeftButtonDown (object sender, MouseEventArgs e)
	{
		can_dragging = true;
	}

	static void HandleMouseLeftButtonUp (object sender, MouseEventArgs e)
	{
		can_dragging = false;
	}

	static void HandleMouseMove (object sender, MouseEventArgs e)
	{
		if (!can_dragging)
			return;

		int x = 0, y = 0;
		window.GetPosition (out x, out y);
		System.Windows.Point p = e.GetPosition (sender as UIElement);
		window.BeginMoveDrag (1, (int) p.X + x,	(int) p.Y + y, 0);

		can_dragging = false;
	}

	static void ConfigureDeskletWindow (Gtk.Window window)
	{
		window.Decorated = false;
		window.SkipPagerHint = true;
		window.SkipTaskbarHint = true;
	}

	static int LoadXaml (string file, List<string> args)
	{
		string [] test = { "" };

		if (sync)
			test [0] = "--sync";

		Application.Init ("mopen", ref test);
		GtkSilver.Init ();
		window = new Window (file);

		string full = System.IO.Path.GetFullPath (file);
		string dir = System.IO.Path.GetDirectoryName (full);
		
		Moonlight.RegisterLoader (delegate (string asm_file) {
			if (System.IO.Path.IsPathRooted (asm_file)){
				return Assembly.LoadFile (asm_file);
			} else {
				return Assembly.LoadFile (System.IO.Path.Combine (dir, asm_file));
			}
		}, delegate (string path) {
			if (System.IO.Path.IsPathRooted (path)){
				return new FileStream (path, FileMode.Open, FileAccess.Read);
			} else {
				return new FileStream (System.IO.Path.Combine (dir, path), FileMode.Open, FileAccess.Read);
			}
		});

		if (transparent) {
			CompositeHelper.SetRgbaColormap (window);
			window.AppPaintable = true;
			window.ExposeEvent += HandleExposeEvent;
		}    	

		if (desklet) {
			ConfigureDeskletWindow (window);
		}

		window.DeleteEvent += delegate {
			Application.Quit ();
		};

		GtkSilver silver = new GtkSilver (400, 400);
		FrameworkElement top;
		
		if (!silver.LoadFile (file, out top)) {
			Console.Error.WriteLine ("mopen: Could not load xaml");
			return 1;
		}

		if (parse_only)
			return 0;

		if (width == -1)
			width = (int) top.Width;
		if (height == -1)
			height = (int) top.Height;

		if (width > 0 && height > 0) {
			silver.Resize (width, height);
			window.Resize (width, height);
		} 

		if (transparent){
			silver.AppPaintable = true;
			silver.Transparent = true;
		}

		if (desklet) {
			top.MouseLeftButtonDown += new MouseButtonEventHandler (HandleMouseLeftButtonDown);
			top.MouseLeftButtonUp += new MouseButtonEventHandler (HandleMouseLeftButtonUp);
			top.MouseMove += new MouseEventHandler (HandleMouseMove);
		}

		window.Add (silver);

		window.ShowAll ();

		if (story_names != null){
			storyboards = new List<Storyboard> ();
			
			foreach (string story in story_names){
				object o = top.FindName (story);
				Storyboard sb = o as Storyboard;

				if (sb == null){
					Console.Error.WriteLine ("mopen: there is no Storyboard object named {0} in the XAML file", story);
					return 1;
				}
				sb.Completed += delegate {
					window.Title = String.Format ("Storyboard {0} completed", current_storyboard-1);
				};
				
				storyboards.Add (sb);
			};

			top.MouseLeftButtonUp += delegate {
				if (current_storyboard == storyboards.Count)
					current_storyboard = 0;
				if (current_storyboard == storyboards.Count)
					return;
				
				window.Title = String.Format ("Storyboard {0} running", current_storyboard);
				storyboards [current_storyboard++].Begin ();
			};
		}

		if (timeout > 0)
			GLib.Timeout.Add ((uint)(timeout * 1000), new TimeoutHandler (Quit));

		Application.Run ();
		return 0;
	}

	static int LoadXap (string file, List<string> args)
	{
		string [] test = { "" };

		if (sync)
			test [0] = "--sync";

		Application.Init ("mopen", ref test);
		GtkSilver.Init ();
		window = new Window (file);

		string full = System.IO.Path.GetFullPath (file);
		string dir = System.IO.Path.GetDirectoryName (full);
		
		Moonlight.RegisterLoader (delegate (string asm_file) {
			if (System.IO.Path.IsPathRooted (asm_file)){
				return Assembly.LoadFile (asm_file);
			} else {
				return Assembly.LoadFile (System.IO.Path.Combine (dir, asm_file));
			}
		}, delegate (string path) {
			if (System.IO.Path.IsPathRooted (path)){
				return new FileStream (path, FileMode.Open, FileAccess.Read);
			} else {
				return new FileStream (System.IO.Path.Combine (dir, path), FileMode.Open, FileAccess.Read);
			}
		});

		if (transparent) {
			CompositeHelper.SetRgbaColormap (window);
			window.AppPaintable = true;
			window.ExposeEvent += HandleExposeEvent;
		}    	

		if (desklet) {
			ConfigureDeskletWindow (window);
		}

		window.DeleteEvent += delegate {
			Application.Quit ();
		};

		GtkSilver silver = new GtkSilver (400, 400);
		System.Windows.Application app;
		
		if (!silver.LoadXap (file, out app)) {
			Console.Error.WriteLine ("mopen: Could not load xaml");
			return 1;
		}

		FrameworkElement top = (FrameworkElement)app.RootVisual;

		if (parse_only)
			return 0;

		if (width == -1)
			width = (int) top.Width;
		if (height == -1)
			height = (int) top.Height;

		if (width > 0 && height > 0) {
			silver.Resize (width, height);
			window.Resize (width, height);
		} 

		if (transparent){
			silver.AppPaintable = true;
			silver.Transparent = true;
		}

		if (desklet) {
			top.MouseLeftButtonDown += new MouseButtonEventHandler (HandleMouseLeftButtonDown);
			top.MouseLeftButtonUp += new MouseButtonEventHandler (HandleMouseLeftButtonUp);
			top.MouseMove += new MouseEventHandler (HandleMouseMove);
		}

		window.Add (silver);

		window.ShowAll ();

		if (story_names != null){
			storyboards = new List<Storyboard> ();
			
			foreach (string story in story_names){
				object o = top.FindName (story);
				Storyboard sb = o as Storyboard;

				if (sb == null){
					Console.Error.WriteLine ("mopen: there is no Storyboard object named {0} in the XAML file", story);
					return 1;
				}
				sb.Completed += delegate {
					window.Title = String.Format ("Storyboard {0} completed", current_storyboard-1);
				};
				
				storyboards.Add (sb);
			};

			top.MouseLeftButtonUp += delegate {
				if (current_storyboard == storyboards.Count)
					current_storyboard = 0;
				if (current_storyboard == storyboards.Count)
					return;
				
				window.Title = String.Format ("Storyboard {0} running", current_storyboard);
				storyboards [current_storyboard++].Begin ();
			};
		}

		if (timeout > 0)
			GLib.Timeout.Add ((uint)(timeout * 1000), new TimeoutHandler (Quit));

		Application.Run ();
		return 0;
	}

	static bool Quit ()
	{
		Application.Quit ();
		return true;
	}

	static bool ParseGeometry (string geometry, out int width, out int height)
	{
		width = 100;
		height = 100;
		
		int p = geometry.IndexOf ('x');
		if (p == -1){
			Console.Error.WriteLine ("Invalid geometry passed: {0}", geometry);
		}
		try {
			width = Int32.Parse (geometry.Substring (0, p));
			height = Int32.Parse (geometry.Substring (p+1));
		} catch {
		}
		
		return true;
	}
	
	//
	// TODO, if a host is specified, look for it, if not,
	// create the domain ourselves and listen to requests.
	//
	static int DoLoad (string file, List<string> args)
	{
		//
		// Here:
		//    implement loading the DLL or executanle, search in path perhaps?
		//
		if (file.EndsWith (".dll"))
			return 1;
		else if (file.EndsWith (".xaml"))
			return LoadXaml (file, args);
		else if (file.EndsWith (".xap"))
			return LoadXap (file, args);

		else
			return 1;
	}
	
	static int Main (string [] args)
	{
		List<string> cmdargs = new List<string> ();
		string [] names;
		string file = null;
		bool help = false;
		
		var p = new OptionSet () {
			{ "h|?|help", v => help = v != null },
			{ "f|fixed", "Disable window resizing", v => fixedwindow = true },
			{ "sync", "Make the gdk connection synchronous", v => sync = true },
			{ "d|desklet", "Remove window decoration for desklets use", v => { desklet = true;  transparent = true; } },
			{ "t|transparent", "Transparent toplevel", v => transparent = true },
			{ "g|geometry=", "Overrides the geometry to be W x H pixels", v => ParseGeometry (v, out width, out height) },
			{ "story=", "Plays the storyboard name N1[,N2,N3], .. Nx when the clicked", v => {
				if (story_names == null)
					story_names = new List<string> ();
				names = v.Split (new Char [] { ','});
				foreach (string s in names)
					story_names.Add (s);
				}
			},
			{ "s|stories", "Automatically prepare to play all stories on click", v => { story_names = new List<string> (); all_stories = true; } },
			{ "parseonly", "Only parse (don't display) the XAML input", v => parse_only = true },
			{ "timeout=", "Time, in seconds, before closing the window", v => {
				if (!Int32.TryParse (v, out timeout)) {
					Console.Error.WriteLine ("mopen: invalid timeout value `{0}'", v);
					Environment.Exit (1);
				}
				}
			},
		};

		List<string> rest;
		try {
			rest = p.Parse (args);
		} catch (OptionException){
			Console.Error.WriteLine ("Try `mxap --help' for more information.");
			return 1;
		}
		
		if (help || rest.Count == 0){
			ShowHelp (p);
			return 1;
		}

		if (rest.Count > 1){
			file = rest [0];
			rest.RemoveAt (0);
			cmdargs = rest;
		} else {
			file = rest [0];
		}

		if (File.Exists (file)) {
			if (all_stories && file.EndsWith (".xaml")) {
				string xns = "http://schemas.microsoft.com/winfx/2006/xaml";
				XmlDocument d = new XmlDocument ();
				d.Load (file);
				
				XmlNamespaceManager mgr = new XmlNamespaceManager (d.NameTable);
				mgr.AddNamespace ("xaml", "http://schemas.microsoft.com/winfx/2006/xaml/presentation");
				mgr.AddNamespace ("x", xns);
				XmlNodeList nodes = d.SelectNodes ("//xaml:Storyboard[@x:Name]", mgr);
				foreach (XmlNode n in nodes){
					XmlAttribute a = n.Attributes ["Name", xns];
					if (a != null)
						story_names.Add (a.Value);
				}
			}
			
			return DoLoad (file, cmdargs);
		}

		if (Directory.Exists (file)){
			string combine = System.IO.Path.Combine (file, "default.xaml");
			if (File.Exists (combine))
				return DoLoad (combine, cmdargs);
		}

		string path = Environment.GetEnvironmentVariable ("PATH");
		string [] dirs = path.Split (new char [] {':'});
		foreach (string dir in dirs){
			string combine = System.IO.Path.Combine (dir, "default.xaml");

			if (File.Exists (combine))
				return DoLoad (combine, cmdargs);
		}
		
		Console.Error.WriteLine ("mopen: Nothing to do");
		return 1;
	}
}
